### DISCUSSION TITLE
Support for OpenAPI

### DISCUSSION THREAD

**Discussion by cpursley** [brainlid/langchain#95] [Ideas]
Something similar to: https://python.langchain.com/docs/integrations/toolkits/openapi

Just learned about this approach here: https://learn.deeplearning.ai/courses/functions-tools-agents-langchain/lesson/6/tools-and-routing

**Comment by brainlid** [comment]
Yes, the Elixir LangChain library already has functions (the only type of tool that exists for OpenAI). I created a `LangChain.Function` and `LangChain.FunctionParam` for defining the tools. I'm working on better explicit support for the "tool" concept though. OpenAI changed their API to be more centered around "tools" instead of functions. See here: https://github.com/brainlid/langchain/issues/48

There's an example of a pre-built tool here: `LangChain.Tools.Calculator`.

There's an example of a router here: `LangChain.Chains.RoutingChain`

As for OpenAPI, I'm not aware of a good Elixir library for creating an OpenAPI client. Do you know one?

**Reply by cpursley** [reply]
Yeah, the existing routing and tools are awesome. I was thinking specifically for utilizing OpenAPI compatible APIs like this: https://medium.com/@jedrzejplucinski/4-steps-to-create-chat-bot-for-your-api-e563c897ef85

Not sure about specific clients, I'll look into it.
